Replace use of deprecated nil_typet in java_type_from_string [blocks: #3800]

tautschnig opened the pull request:
Use typet or optionalt<typet>. ~~Do not review just yet, I will split this up in several commit as it touches way too many parts of the codebase in a single commit.~~

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Replace use of deprecated nil_typet in java_type_from_string

Use optionalt<typet> as recommended in the deprecation note.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4aad8d183a3939bb7f0c263126148b77f0130def).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98657901
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 152dfcc2932b6914411b272a765a1799ce1a0e87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98710526
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

peterschrammel commented on jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp:
```diff
@@ -240,11 +240,11 @@ SCENARIO(
           symbol_exprt generic_param{
             "specalisedGeneric",
-            java_type_from_string("LDummyGeneric<Ljava/lang/Interger;>;")};
+            *java_type_from_string("LDummyGeneric<Ljava/lang/Interger;>;")};
```
❓ Is `Interger` intended here (and below)?

tautschnig replied:
I guess it doesn't matter? db6756e5 - @thk123 ?

thk123 replied:
I guess not...! Will have a detailed look at this test, but of course no more or less broken than it was before.

peterschrammel commented on jbmc/src/java_bytecode/java_types.cpp:
```diff
@@ -582,26 +582,25 @@ typet java_type_from_string(
         [](const typet &type) { return java_method_typet::parametert(type); });
 
-      return java_method_typet(std::move(parameters), std::move(return_type));
+      return java_method_typet(std::move(parameters), std::move(*return_type));
```
Are we sure this has a value?

tautschnig replied:
I have no idea, but it was previously used without any sanity check, so I'd now rely on `operator*` failing an assertion...

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -637,10 +637,8 @@ void java_bytecode_convert_classt::convert(
-      f.descriptor,
-      f.signature,
-      id2string(class_symbol.name));
+    field_type = *java_type_from_string_with_exception(
```
Are we sure this has a value? (and other places)

tautschnig replied:
As above: no idea, but it was previously used unchecked. There are a couple of places that failed this while running the regression tests, and those I have now guarded appropriately. I still think that this makes `optionalt<typet>` much better as at least you'd have a failing assertion instead of silently passing around a `nil_typet` that some later stage won't know how to deal with.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -460,24 +465,10 @@ void java_bytecode_convert_methodt::convert(
     id_oss << method_id << "::" << v.name;
     irep_idt identifier(id_oss.str());
-    symbol_exprt result(identifier, t);
+    symbol_exprt result = symbol_exprt::typeless(identifier);
```
:question: This change looks a little suspicious? Previously the type of the symbol was derived from the signature or descriptor, now is a typeless identifier?

tautschnig replied:
The type was sometimes `nil_typet()` (because trying to derive the type failed) and nobody noticed, because really the type of this is never used.

smowton replied:
In that case could I suggest using an `irep_idt` instead of a `symbol_exprt`? Otherwise someone will surely assume that `variables[n].type()` is a sensible thing to look at.

tautschnig replied:
That would necessitate some refactoring of this code as the expression is also used to communicate `ID_C_base_name`.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -548,8 +548,8 @@ void java_bytecode_parsert::get_class_refs()
       {
-        typet t=java_type_from_string(id2string(pool_entry(c.ref2).s));
-        get_class_refs_rec(t);
+        const auto t = java_type_from_string(id2string(pool_entry(c.ref2).s));
```
:pick: while in the area, `t` -> `type`

tautschnig replied:
I just got rid of it, it didn't carry much value...

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -562,59 +562,57 @@ void java_bytecode_parsert::get_class_refs()
-    else
-      method_type=java_type_from_string(method.descriptor);
+    else if(const auto method_type = java_type_from_string(method.descriptor))
+      get_class_refs_rec(*method_type);
```
:bulb: Might be worth sticking a warning here that it somehow failed to get the type from both the signature and the descriptor. I'd even be tempted to put an `UNREACHABLE;` here, but I suppose this is one of the cases you were referring to where this broke tests.

tautschnig replied:
Indeed, this is reachable/reached in several regression tests. I don't feel sufficiently competent on this code to assess whether that's actually a problem. Any volunteers?

thk123 replied:
I put `UNREACHABLE;` here and didn't get any failures in `jbmc*` - have a specific example? (will try the other two places too).

tautschnig replied:
Not sure what exactly your change was, but I'd just use `else { typet method_type = *java_type_from_string(method.descriptor); get_class_refs_rec(method_type); }` here.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -562,59 +562,57 @@ void java_bytecode_parsert::get_class_refs()
-        var_type=java_type_from_string(var.descriptor);
-      get_class_refs_rec(var_type);
+      else if(const auto var_type = java_type_from_string(var.descriptor))
+        get_class_refs_rec(*var_type);
```
:bulb: Again - some kind of warning in the `else` case makes sense here I think. :fire:

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -672,8 +670,9 @@ void java_bytecode_parsert::get_annotation_value_class_refs(const exprt &value)
-    get_class_refs_rec(value_type);
+    const auto value_type = java_type_from_string(id2string(value_id));
+    if(value_type.has_value())
+      get_class_refs_rec(*value_type);
```
:bulb: :fire:

thk123 reviewed (commented):
Will create a manual bump for this.

tautschnig commented:
I have rebased this one. @thk123 or anyone else (@smowton ?) who feels comfortable with the Java bytecode parser: the cases where I'm now silently skipping calls to `get_class_refs_rec` may be worth exploring as highlighted by @thk123 above. They are easily triggered via the regression test suite.

thk123 commented:
I will have a look and see whether this looks like a real concern or not.

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -259,21 +259,26 @@ java_method_typet member_type_lazy(
-      INVARIANT(member_type_from_signature.id()==ID_code, "Must be code type");
+      auto member_type_from_signature =
+        java_type_from_string(signature.value(), class_name);
+      INVARIANT(
```
I'd actually delete this, we're about to check both those things by deref'ing and casting

smowton commented:
@tautschnig it's unsurprising with our known-to-be-broken generic signature parsing sometimes bails with `nil_typet` (we should probably treat that the same as the exception thrown when we find something we can't handle and fall back to the descriptor). If that happens with a descriptor (non-generic type) then I'm pretty surprised.

thk123 reviewed (changes requested):
This breaks TG (both a trivial compile error and a failing test) I'm investigating now.

thk123 commented:
Now waiting on CI - in the mean time will investigate the above optional returns.

tautschnig commented:
Thank you @thk123 !

Commit:
Address errors in handcrafted bytecode

The jasmin contained two errors:
calling base class constructor with invokevirtual rather than
invokespecial
the descriptor of the type for the local variable was not a valid
descriptor.

Commit:
Do not allow getting type from descriptor to fail

There should be no unsupported cases here.

thk123 commented:
@tautschnig I've made tautschnig/cbmc#12 which tightens these to not allow the descriptor to fail (I could only find one failing test - and the problem was the test). Happy for you to fold that into this PR or get this one merged and then I'll raise that as a new PR. 

(TG bump is based at this PR and is still running, but passed some tests locally).

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: c523d5f6a1878a41824d9a6b741bf1415b3c448a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99277071
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented:
@thk123 PR merged in here and the entire branch rebased onto latest develop, you might want to update the TG bump to use this?

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 8f71a23c0ba06cedd82e49bcd84a9925352e4ac8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99300653
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thk123 commented:
TG bump passing.
